### DISCUSSION TITLE
⚡️ Android - Restart `Screen Capture` service on boot

### DISCUSSION THREAD

**Discussion by Rikj000** [rustdesk/rustdesk#1330] [General]
## Issue Description

Currently, when using RustDesk for remote customer support on Android,   
there is no easy way to automatically re-enable the `Screen Capture` service on boot.

Which means that the customer needs to be contacted & informed,   
on how to re-enable the `Screen Capture` service when support is needed.   
This slows down the support process / makes it less convenient for both parties.

##  Workaround
- Android => RustDesk => Share Screen => Permissions => Enable `Screen capture`
- Android => RustDesk => Share Screen => Permissions => Enable `Input control`
- Android => RustDesk => Share Screen => Permissions => Enable `Transfer file`
- Android => RustDesk => Share Screen => Permissions => Enable `Audio capture`
- Android => RustDesk => Share Screen => Triple dot icon => Select `Use permanent password`
- Android => RustDesk => Share Screen => Triple dot icon => Select `Set permanent password`
- Android => RustDesk => Settings => Enhancements => Enable `Keep RustDesk background service`
- Android => RustDesk => Settings => Enhancements => Enable `Start on boot (beta)` 
    => Enable `Permit drawing over other apps`
- Desktop => Terminal => Connect with ADB and run:
    ```bash
    adb shell appops set com.carriez.flutter_hbb PROJECT_MEDIA allow
    ```
    [Setting AppOps](https://community.teamviewer.com/English/discussion/105147/have-to-allow-each-time-with-host-app) is required to fully automatically start RustDesk on boot on Android 13

https://www.reddit.com/r/AnyDesk/comments/14m054k/comment/m0h4y8y/?utm_source=share&utm_medium=web3x&utm_name=web3xcss&utm_term=1&utm_content=share_button

**Comment by ytoaa** [comment]
Did you happen to get the automate script?

**Reply by vzb007** [reply]
i am not a programmer. but we can see the implementation of droid-ng vnc. they have a very good implementation of this issue.  this would make rustdesk very much the defacto software as it kicks arse with the features. look at the end of the commands.




Remote Control via the Intent Interface
droidVNC-NG features a remote control interface by means of Intents. This allows starting the VNC server from other apps or on certain events. It is designed to be working with automation apps like [MacroDroid](https://www.macrodroid.com/), [Automate](https://llamalab.com/automate/) or [Tasker](https://tasker.joaoapps.com/) as well as to be called from code.

You basically send an explicit Intent to net.christianbeier.droidvnc_ng.MainService with one of the following Actions and associated Extras set:

net.christianbeier.droidvnc_ng.ACTION_START: Starts the server.

net.christianbeier.droidvnc_ng.EXTRA_ACCESS_KEY: Required String Extra containing the remote control interface's access key. You can get/set this from the Admin Panel.
net.christianbeier.droidvnc_ng.EXTRA_REQUEST_ID: Optional String Extra containing a unique id for this request. Used to identify the answer from the service.
net.christianbeier.droidvnc_ng.EXTRA_PORT: Optional Integer Extra setting the listening port. Set to -1 to disable listening.
net.christianbeier.droidvnc_ng.EXTRA_PASSWORD: Optional String Extra containing VNC password.
net.christianbeier.droidvnc_ng.EXTRA_SCALING: Optional Float Extra between 0.0 and 1.0 describing the server-side framebuffer scaling.
net.christianbeier.droidvnc_ng.EXTRA_VIEW_ONLY: Optional Boolean Extra toggling view-only mode.
net.christianbeier.droidvnc_ng.EXTRA_SHOW_POINTERS: Optional Boolean Extra toggling per-client mouse pointers.
net.christianbeier.droidvnc_ng.EXTRA_FILE_TRANSFER: Optional Boolean Extra toggling the file transfer feature.
net.christianbeier.droidvnc_ng.ACTION_CONNECT_REVERSE: Make an outbound connection to a listening viewer.

net.christianbeier.droidvnc_ng.EXTRA_ACCESS_KEY: Required String Extra containing the remote control interface's access key. You can get/set this from the Admin Panel.
net.christianbeier.droidvnc_ng.EXTRA_REQUEST_ID: Optional String Extra containing a unique id for this request. Used to identify the answer from the service.
net.christianbeier.droidvnc_ng.EXTRA_HOST: Required String Extra setting the host to connect to.
net.christianbeier.droidvnc_ng.EXTRA_PORT: Optional Integer Extra setting the remote port.
net.christianbeier.droidvnc_ng.ACTION_CONNECT_REPEATER Make an outbound connection to a repeater.

net.christianbeier.droidvnc_ng.EXTRA_ACCESS_KEY: Required String Extra containing the remote control interface's access key. You can get/set this from the Admin Panel.
net.christianbeier.droidvnc_ng.EXTRA_REQUEST_ID: Optional String Extra containing a unique id for this request. Used to identify the answer from the service.
net.christianbeier.droidvnc_ng.EXTRA_HOST: Required String Extra setting the host to connect to.
net.christianbeier.droidvnc_ng.EXTRA_PORT: Optional Integer Extra setting the remote port.
net.christianbeier.droidvnc_ng.EXTRA_REPEATER_ID: Required String Extra setting the ID on the repeater.
net.christianbeier.droidvnc_ng.ACTION_STOP: Stops the server.

net.christianbeier.droidvnc_ng.EXTRA_ACCESS_KEY: Required String Extra containing the remote control interface's access key. You can get/set this from the Admin Panel.
net.christianbeier.droidvnc_ng.EXTRA_REQUEST_ID: Optional String Extra containing a unique id for this request. Used to identify the answer from the service.
The service answers with a Broadcast Intent with its Action mirroring your request:

Action: one of the above Actions you requested
net.christianbeier.droidvnc_ng.EXTRA_REQUEST_ID: The request id this answer is for.
net.christianbeier.droidvnc_ng.EXTRA_REQUEST_SUCCESS: Boolean Extra describing the outcome of the request.
There is one special case where the service sends a Broadcast Intent with action net.christianbeier.droidvnc_ng.ACTION_STOP without any extras: that is when it is stopped by the system.

Examples
Start a password-protected view-only server on port 5901
Using adb shell am syntax:

adb shell am start-foreground-service \
 -n net.christianbeier.droidvnc_ng/.MainService \
 -a net.christianbeier.droidvnc_ng.ACTION_START \
 --es net.christianbeier.droidvnc_ng.EXTRA_ACCESS_KEY de32550a6efb43f8a5d145e6c07b2cde \
 --es net.christianbeier.droidvnc_ng.EXTRA_REQUEST_ID abc123 \
 --ei net.christianbeier.droidvnc_ng.EXTRA_PORT 5901 \
 --es net.christianbeier.droidvnc_ng.EXTRA_PASSWORD supersecure \
 --ez net.christianbeier.droidvnc_ng.EXTRA_VIEW_ONLY true
Make an outbound connection to a listening viewer from the running server
For example from Java code:

See [MainActivity.java](https://github.com/bk138/droidVNC-NG/blob/master/app/src/main/java/net/christianbeier/droidvnc_ng/MainActivity.java).

Stop the server again
Using adb shell am syntax again:

adb shell am start-foreground-service \
 -n net.christianbeier.droidvnc_ng/.MainService \
 -a net.christianbeier.droidvnc_ng.ACTION_STOP \
 --es net.christianbeier.droidvnc_ng.EXTRA_ACCESS_KEY de32550a6efb43f8a5d145e6c07b2cde \
 --es net.christianbeier.droidvnc_ng.EXTRA_REQUEST_ID def456


Notes
Requires at least Android 7.

**[Since Android 10](https://developer.android.com/about/versions/10/privacy/changes#screen-contents), the permission to access the screen contents has to be given on each start and is not saved. You can, however, work around this by installing [adb](https://developer.android.com/studio/command-line/adb) (or simply Android Studio) on a PC, connecting the device running droidVNC-NG to that PC and running adb shell cmd appops set net.christianbeier.droidvnc_ng PROJECT_MEDIA allow once.

You can also use adb to manually give input permission prior to app start via adb shell settings put secure enabled_accessibility_services net.christianbeier.droidvnc_ng/.InputService:$(adb shell settings get secure enabled_accessibility_services).

If you are getting a black screen in a connected VNC viewer despite having given all permissions, it might be that your device does not support Android's MediaProjection API correctly. To find out, you can try screen recording with another app, [ScreenRecorder](https://gitlab.com/vijai/screenrecorder). If it fails as well, your device most likely does not support screen recording via MediaProjection. This is known to be the case for [Android-x86](https://www.android-x86.org/).**



hope this helps

**Comment by rustdesk** [comment]
> **[Since Android 10](https://developer.android.com/about/versions/10/privacy/changes#screen-contents), the permission to access the screen contents has to be given on each start and is not saved. You can, however, work around this by installing [adb](https://developer.android.com/studio/command-line/adb) (or simply Android Studio) on a PC, connecting the device running droidVNC-NG to that PC and running adb shell cmd appops set net.christianbeier.droidvnc_ng PROJECT_MEDIA allow once.
> 
> You can also use adb to manually give input permission prior to app start via adb shell settings put secure enabled_accessibility_services net.christianbeier.droidvnc_ng/.InputService:$(adb shell settings get secure enabled_accessibility_services).
> 
> If you are getting a black screen in a connected VNC viewer despite having given all permissions, it might be that your device does not support Android's MediaProjection API correctly. To find out, you can try screen recording with another app, [ScreenRecorder](https://gitlab.com/vijai/screenrecorder). If it fails as well, your device most likely does not support screen recording via MediaProjection. This is known to be the case for [Android-x86](https://www.android-x86.org/).**

👍

**Comment by TheGrave** [comment]
There is a workaround with Shizuku and AppOps that works for all remote control apps:

https://community.teamviewer.com/English/discussion/105147/have-to-allow-each-time-with-host-app

**Reply by Rikj000** [reply]
Shizuku requires root access though.

Which may be suitable for individual power users.
   
However it's not suitable for production/customer devices,   
since it imposes security risks.

**Reply by TheGrave** [reply]
BS, it requires USB or wireless debugging only.

**Reply by dixtdf** [reply]
Using Magisk+Sui+AppOps is very useful. Thank you for your reminder, and I hope I can remember it in the future

**Comment by Chen017** [comment]
yes，service cannot auto start when the app is killed except reboot，which means i need open it manually

**Comment by vcheckzen** [comment]
The following command works for me on android 13 with no root.

```
adb shell appops set com.carriez.flutter_hbb PROJECT_MEDIA allow
```

Ref <https://tasker.joaoapps.com/userguide/en/help/ah_screencapture_grant.html>

**Comment by zekuraz** [comment]
Hi !

We have a fleet of Viewsonic CDE5520 and CDE4320 screens.

We'd like to control these screens remotely. To do this, we have installed the RustDesk Client software (https://rustdesk.com/docs/en/client/). The software works. However, to save energy, we have set up a system that puts the screens to sleep. As soon as the screens are restarted, the RustDesk Client software services do not restart automatically. So you have to move around manually and restart the software, which is annoying.

I noticed that the application has a "Start on Boot" feature that requires special permissions (see attachment). However, when I click on it, nothing happens. I wanted to know if there is a procedure for enabling these permissions on android.

(Another remark, we also have CDE6520, and on this model it works perfectly with the restart of the RustDesk Client service. I access the special permissions on this model)

thank you in advance for your clarifications

**Reply by Rikj000** [reply]
Try setting the appops as described by @vcheckzen in the post above yours: https://github.com/rustdesk/rustdesk/discussions/1330#discussioncomment-8590114

**Comment by Rikj000** [comment]
Closing discussion as resolved, updated [Workaround](https://github.com/rustdesk/rustdesk/discussions/1330#discussion-4325565)
